### PR TITLE
chore: extract navigation tracking

### PR DESCRIPTION
**Pull request by sadym-chromium**
Extract navigation tracking logic to a dedicated class. Simplify the logic. Required to be able to make further adjustments to the navigation logic.

Out of scope: document tracking.

### CODE REVIEW

File: src/bidiMapper/modules/context/NavigationTracker.ts [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+  // Tracks CDP events on the navigation. Used to distinguish new navigations from the
+  // initial one.
+  cdpStates = {
+    frameScheduledNavigation: false,
+    frameRequestedNavigation: false,
+    frameStartedLoading: false,
+  };
+
+  markFragmentNavigated() {
+    this.started.resolve();
```
Comment (OrKoN) [inline src/bidiMapper/modules/context/NavigationTracker.ts:49]: so fragment navigations should not have other navigation events like navigationStarted as per spec.

File: src/bidiMapper/modules/context/NavigationTracker.ts [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+      url: this.url,
+    };
+  }
+
+  markStarted(): void {
+    this.started.resolve();
+  }
+}
+
+export class NavigationTracker {
```
Comment (OrKoN) [inline src/bidiMapper/modules/context/NavigationTracker.ts:92]: could we perhaps include unit tests to indicate all the expectations we have from the NavigationTracker (similar to what Nik has done for the network manager/storage)

File: src/bidiMapper/modules/context/NavigationTracker.ts [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+  frameRequestedNavigation = (url: string): void => {
+    // Signals that it's a new navigation:
+    // 1. The `Page.frameStartedLoading` was already emitted.
+    // 2. The `Page.frameRequestedNavigation` was already emitted. Note that having
+    //    `Page.frameScheduledNavigation` emitted at this point does not signal the new
+    //    navigation.
+    // 3. The current navigation is the initial one, and the URL does not match
+    //    `about:blank`.
+    const isNewNavigation =
+      this.#currentNavigation.cdpStates.frameStartedLoading ||
```
Comment (OrKoN) [inline src/bidiMapper/modules/context/NavigationTracker.ts:202]: how does cdpStates track if the events came from a current navigation or from one of the past navigations?

**Review by OrKoN — APPROVED** [review]
It's a little difficult to track if the behavior is the same. Assuming it is, this LGTM. I would suggest adding unit tests for various navigation tracker behaviors.

File: src/utils/Time.spec.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
+/**
```
Comment (Lightning00Blade) [inline src/utils/Time.spec.ts:1]: Nit: non Class files should start with small letter.

File: src/bidiMapper/modules/context/BrowsingContextImpl.ts [hunk 5]
```diff
@@ -419,8 +391,9 @@ export class BrowsingContextImpl {
       if (this.id !== params.frame.id) {
         return;
       }
-      this.#url = params.frame.url + (params.frame.urlFragment ?? '');
-      this.#pendingNavigationUrl = undefined;
+      const url = params.frame.url + (params.frame.urlFragment ?? '');
+      this.#url = url;
```
Comment (Lightning00Blade) [inline src/bidiMapper/modules/context/BrowsingContextImpl.ts:395]: In a follow up would it make sense to refactor the Url to come from the navigation tracking as it knows when its committed and when not.
So we change 
```ts
get #url(){
  return this.#navigationTracker.url;
}
```

Comment (sadym-chromium) [reply src/bidiMapper/modules/context/BrowsingContextImpl.ts:395]: I guess it should come from document tracker instead when it is implemented.

File: src/bidiMapper/modules/context/NavigationTracker.ts (hunk 2 above)
Comment (Lightning00Blade) [reply src/bidiMapper/modules/context/NavigationTracker.ts:92]: +1 to this.

**Review by Lightning00Blade — APPROVED** [review]

**Comment by sadym-chromium** [comment]
closing in favor of https://github.com/GoogleChromeLabs/chromium-bidi/pull/2881

**Pull request closed without merging**